### PR TITLE
Trigger export to Hubble and export assets

### PR DESCRIPTION
**Pull request by robbedec**
[Link to documentation](https://www.notion.so/Adobe-XD-b67717dbef514c139c557f0f60400b93#4f3df7b643d948e9aba814627fde27cb)

Possible result of asset export

Schermafbeelding 2020-05-06 om 08 23 20

### CODE REVIEW

File: main.js [hunk 1]
```diff
@@ -0,0 +1,9 @@
+const exportToHubble = require('./src/exportToHubble');
+const exportAssets = require('./src/exportAssets');
+
+module.exports = {
+    commands: {
+        exportToHubble: exportToHubble,
+        exportAssets: exportAssets,
```
Comment (thibmaek) [inline main.js]:
```suggestion
        exportToHubble,
        exportAssets,
```

File: manifest.json [hunk 2]
```diff
@@ -0,0 +1,33 @@
+{
+  "name": "hubble",
+  "host": {
+    "app": "XD",
+    "minVersion": "21.0"
+  },
+  "id" : "6d286a66",
```
Comment (thibmaek) [inline manifest.json:7]: This is just random?

File: manifest.json [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+      "type": "menu",
+      "commandId": "exportToHubble"
+    },
+    {
+      "label": "Export assets",
+      "type": "menu",
+      "commandId": "exportAssets"
+    }
+  ],
+  "version": "0.0.1"
```
Comment (thibmaek) [inline manifest.json]:
```suggestion
  "version": "1.0.0"
```

File: src/exportAssets.js [hunk 4]
```diff
@@ -0,0 +1,55 @@
+const fs = require('uxp').storage.localFileSystem;
```
Comment (thibmaek) [inline src/exportAssets.js:1]: 🤮

File: src/exportAssets.js [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    } catch(err) {
+      folder = await folder.createFolder(name);
+    }
+  }
+
+  return folder;
+}
+
+async function exportAssets(_selection, documentRoot) {
+  // TODO: confirm if marked_for export is enough or if every icon/illustration should be exported
```
Comment (thibmaek) [inline src/exportAssets.js]: Only marked for export, Sketch and Figma are handled the same

File: src/exportAssets.js [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+   * Run sequential because application.createRenditions is not concurrent
+   * so we have to wait until the previous has resolved before starting a new one
+   */
+   for (const asset of assets) {
+    const path = asset.name.split('/');
+
+    const folderWithFile = await createDirectoryPath(folder, path);
+    const renderFile = await folderWithFile.createFile(`${path.pop()}.png`, { overwrite: true });
+
+    // TODO: check which output formats are necessary
```
Comment (thibmaek) [inline src/exportAssets.js]: Preference is: if we can get it from the Adobe XD file (e.g designer marks 1,2,3x png and SVG) to reuse that. If we can not get that from the file (like in Sketch) we use 1,2,3x png and SVG hardcoded

File: src/exportAssets.js [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+  const folder = await fs.getFolder();
+  if (!folder) return;
+
+  /**
+   * Run sequential because application.createRenditions is not concurrent
+   * so we have to wait until the previous has resolved before starting a new one
+   */
+   for (const asset of assets) {
+    const path = asset.name.split('/');
```
Comment (thibmaek) [inline src/exportAssets.js:30]: If you can use Node's `path` module you could go for `path.basename(asset.name)` here.
A basename is the lowest child (e.g file name) inside a tree of folders with its extension

File: src/exportAssets.js [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+      {
+        node: asset,
+        outputFile: renderFile,
+        type: application.RenditionType.PNG,
+        scale: 2
+      }
+    ];
+
+    try {
+      await application.createRenditions(renditionOptions);
```
Comment (thibmaek) [inline src/exportAssets.js]: Don't the other promises have to be inside a try-catch block as well? If one of those promises throws, the exception would be unhandled. Why not wrap everything async in try-catch?

Remember: try-catch is just a different way of saying `promise.catch()` for async await

File: src/exportToHubble.js [hunk 9]
```diff
@@ -0,0 +1,83 @@
+const fs = require('uxp').storage.localFileSystem;
+const shell = require('uxp').shell;
```
Comment (thibmaek) [inline src/exportToHubble.js]:
```suggestion
const { shell } = require('uxp');
```

File: src/exportToHubble.js [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+
+    const dialog = document.getElementById('dialog');
+
+    const cancelButton = document.getElementById('cancel');
+    cancelButton.addEventListener('click', () => dialog.close(CANCELLED));
+
+    const okButton = document.getElementById('ok');
+    okButton.addEventListener('click', e => {
+        dialog.close(document.getElementById('url').value);
+        e.preventDefault();
```
Comment (thibmaek) [inline src/exportToHubble.js:53]: I'd call cancellation of event before performing any side-effects myself.

File: src/exportToHubble.js [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+    return dialog.showModal();
+}
+
+async function exportToHubble() { 
+    const file = await fs.getFileForOpening({types:['xd']});
+    if (!file) return;
+    const filePath = file.nativePath;
+
+    const folder = await fs.getDataFolder();
+    let urlFile = null;
```
Comment (thibmaek) [inline src/exportToHubble.js]:
```suggestion
    let urlFile;
```

File: src/exportToHubble.js [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+    try {
+        urlFile = await folder.getEntry('cloudurl.txt');
```
Comment (thibmaek) [inline src/exportToHubble.js:68]: Wrapping this in a try-catch doesn't make sense if you are just going to repeat it again in the catch.

If this throws inside the first attempt, why would you reexecute it in the catch block, just to have it throw again?

**Review by thibmaek — CHANGES_REQUESTED** [review]

File: manifest.json (hunk 2 above)
Comment (robbedec) [reply manifest.json:7]: Generated from the adobe console

File: src/exportToHubble.js (hunk 12 above)
Comment (robbedec) [reply src/exportToHubble.js:68]: Those are 2 different functions. The initial call tries to get the file from the data folder. If the call fails (i.e. it doesn't exist) we create the file in the catch block

Comment (thibmaek) [reply src/exportToHubble.js:68]: Haha wow wasn't paying attention

File: src/exportAssets.js (hunk 7 above)
Comment (robbedec) [reply src/exportAssets.js:30]: The individual parts are used to create the directory structure anyways and I just checked, `path` is not available

File: src/exportAssets.js [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+   for (const asset of assets) {
+    const path = asset.name.split('/');
+
+    const folderWithFile = await createDirectoryPath(folder, path);
+    const fileName = path.pop();
+
+    for (const scale of [1, 2, 3]) {
+      try {
+        const renderFilePNG = await folderWithFile.createFile(`${fileName}@${scale}x.png`, { overwrite: true });
+        const renderFileSVG = await folderWithFile.createFile(`${fileName}@${scale}x.svg`, { overwrite: true });
```
Comment (thibmaek) [inline src/exportAssets.js]: SVGs should only be rendered in 1 dimension

File: src/exportToHubble.js [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+    } catch (err) {
+        urlFile = await folder.createFile('cloudurl.txt');
+        // Something has to be written to the file before it shows up in the filesystem
+        await urlFile.write('');
+    }
+    const defaultInput = await urlFile.read();
+    const url = await showDialog(defaultInput);
+
+    if (url != CANCELLED) {
```
Comment (thibmaek) [inline src/exportToHubble.js]: !== ?

**Review by thibmaek — CHANGES_REQUESTED** [review]

File: src/exportToHubble.js (hunk 14 above)
Comment (robbedec) [reply src/exportToHubble.js]: Woops

File: src/exportAssets.js (hunk 13 above)
Comment (robbedec) [reply src/exportAssets.js]: Dimension, as in scale?

**Review by thibmaek — APPROVED** [review]

**Comment by thibmaek** [comment]
Can you (maybe in a next PR) update the README, like the examples for Sketch and Figma?

**Pull request merged**